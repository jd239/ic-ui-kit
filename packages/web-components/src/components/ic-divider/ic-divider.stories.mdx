import { Meta, Story, Canvas, Description } from "@storybook/addon-docs";
import { Markdown } from "@storybook/blocks";
import { html } from "lit-html";
import readme from "./readme.md?raw";

<Meta
  title="Web Components/Divider"
  component="ic-divider"
  parameters={{ loki: { skip: true } }}
/>

<Markdown>{readme}</Markdown>

### Theme

<Canvas>
  <Story name="Theme">
    {html` <div
      style="display: flex; flex-direction: column; gap: var(--ic-space-md)"
    >
      <ic-divider
        theme="inherit"
        weight="medium"
        label="theme='inherit'"
        label-placement="left"
      ></ic-divider>
      <ic-divider
        theme="inherit"
        monochrome="true"
        weight="medium"
        label="theme='inherit', monochrome='true'"
        label-placement="left"
      ></ic-divider>
      <div
        style="display:flex; flex-direction: column; align-items: center; background-color: #fff; width: 100%; padding: var(--ic-space-xs) 0; gap: var(--ic-space-md)"
      >
        <ic-divider
          theme="light"
          weight="medium"
          label="theme='light'"
          label-placement="left"
        ></ic-divider>
        <ic-divider
          theme="light"
          monochrome="true"
          weight="medium"
          label="theme='light', monochrome='true'"
          label-placement="left"
        ></ic-divider>
      </div>
      <div
        style="display:flex; flex-direction: column; align-items: center; background-color: var(--ic-color-background-primary-dark); width: 100%; padding: var(--ic-space-xs) 0; gap: var(--ic-space-md)"
      >
        <ic-divider
          theme="dark"
          weight="medium"
          label="theme='dark'"
          label-placement="left"
        ></ic-divider>
        <ic-divider
          theme="dark"
          monochrome="true"
          weight="medium"
          label="theme='dark', monochrome='true'"
          label-placement="left"
        ></ic-divider>
      </div>
    </div>`}
  </Story>
</Canvas>

### Weight

<Canvas>
  <Story name="Weight">
    {html` <div
      style="display: flex; flex-direction: column; gap: var(--ic-space-md)"
    >
      <ic-divider weight="thin"></ic-divider>
      <ic-divider weight="thin" border-style="dashed"></ic-divider>
      <ic-divider
        weight="thin"
        label="End of section"
        label-placement="center"
      ></ic-divider>
      <ic-divider weight="medium"></ic-divider>
      <ic-divider weight="medium" border-style="dashed"></ic-divider>
      <ic-divider
        weight="medium"
        label="End of section"
        label-placement="center"
      ></ic-divider>
      <ic-divider weight="thick"></ic-divider>
      <ic-divider weight="thick" border-style="dashed"></ic-divider>
      <ic-divider
        weight="thick"
        label="End of section"
        label-placement="center"
      ></ic-divider>
      <ic-divider weight="very-thick"></ic-divider>
      <ic-divider weight="very-thick" border-style="dashed"></ic-divider>
      <ic-divider
        weight="very-thick"
        label="End of section"
        label-placement="center"
      ></ic-divider>
    </div>`}
  </Story>
</Canvas>

### Border style

<Canvas>
  <Story name="Border style">
    {html` <div
      style="display: flex; flex-direction: column; gap: var(--ic-space-md)"
    >
      <ic-divider border-style="solid"></ic-divider>
      <ic-divider border-style="dashed"></ic-divider>
    </div>`}
  </Story>
</Canvas>

### Vertical

<Canvas>
  <Story name="Vertical">
    {html`
      <div style="display: flex; gap: var(--ic-space-lg); height: 400px">
        <ic-divider label-placement="top" orientation="vertical" weight="thin">
          <ic-typography variant="label" slot="label">
            <p style="width: max-content;">Slotted label top</p>
          </ic-typography></ic-divider
        >
        <ic-divider
          orientation="vertical"
          weight="thin"
          border-style="dashed"
        ></ic-divider>
        <ic-divider orientation="vertical" weight="medium"></ic-divider>
        <ic-divider
          label="Label center"
          label-placement="center"
          orientation="vertical"
          border-style="dashed"
          weight="medium"
        ></ic-divider>
        <ic-divider orientation="vertical" weight="thick"></ic-divider>
        <ic-divider
          orientation="vertical"
          weight="thick"
          border-style="dashed"
        ></ic-divider>
        <ic-divider orientation="vertical" weight="very-thick"></ic-divider>
        <ic-divider
          orientation="vertical"
          weight="very-thick"
          border-style="dashed"
        ></ic-divider>
      </div>
    `}
  </Story>
</Canvas>

### Max content

<Canvas>
  <Story name="Max content">
  {html` <div
        style="display: flex; flex-direction: column; gap: var(--ic-space-lg)"
      >
        <ic-divider 
          label="ðŸŽ¶ Now this is a story all about how my life got brewed up and turned around. And Iâ€™d like to take a minute, just sit right there; Iâ€™ll tell you how I became the prince of a cafÃ© called Bel-Air ðŸŽ¶
          ðŸŽ¶ In west Philadelphia, born and raised; in the coffee shop was where I spent most of my days. Chillinâ€™ out, maxinâ€™, relaxinâ€™ all cool, sippinâ€™ on some cappuccinos outside of the school ðŸŽ¶"
          label-placement="left"></ic-divider
        >
        <ic-divider 
          label="ðŸŽ¶ Now this is a story all about how my life got brewed up and turned around. And Iâ€™d like to take a minute, just sit right there; Iâ€™ll tell you how I became the prince of a cafÃ© called Bel-Air ðŸŽ¶
          ðŸŽ¶ In west Philadelphia, born and raised; in the coffee shop was where I spent most of my days. Chillinâ€™ out, maxinâ€™, relaxinâ€™ all cool, sippinâ€™ on some cappuccinos outside of the school ðŸŽ¶"
          label-placement="center"></ic-divider
        >
      </div>
      <ic-divider label-placement="right">
          <ic-typography variant="label" slot="label" style="width: max-content;max-width: 100%;">
            <p style="width: inherit;">A slotted label: ðŸŽ¶ Now this is a story all about how my life got brewed up and turned around. And Iâ€™d like to take a minute, just sit right there; Iâ€™ll tell you how I became the prince of a cafÃ© called Bel-Air ðŸŽ¶
        ðŸŽ¶ In west Philadelphia, born and raised; in the coffee shop was where I spent most of my days. Chillinâ€™ out, maxinâ€™, relaxinâ€™ all cool, sippinâ€™ on some cappuccinos outside of the school ðŸŽ¶</p>
          </ic-typography></ic-divider
        >
      </div>
      <div
        style="display: flex; gap: var(--ic-space-lg); height: 600px; margin-top: var(--ic-space-lg)"
      >
        <ic-divider
          label="A vertical divider: ðŸŽ¶ Now this is a story all about how my life got brewed up and turned around. And Iâ€™d like to take a minute, just sit right there; Iâ€™ll tell you how I became the prince of a cafÃ© called Bel-Air ðŸŽ¶
        ðŸŽ¶ In west Philadelphia, born and raised; in the coffee shop was where I spent most of my days. Chillinâ€™ out, maxinâ€™, relaxinâ€™ all cool, sippinâ€™ on some cappuccinos outside of the school ðŸŽ¶"
          label-placement="top"
          orientation="vertical"
        ></ic-divider>
      </div>`}

  </Story>
</Canvas>

### Label placement

<Canvas>
  <Story name="Label placement">
    {html` <div
        style="display: flex; flex-direction: column; gap: var(--ic-space-lg)"
      >
        <ic-divider label="Label left" label-placement="left"></ic-divider>
        <ic-divider label="Label center" label-placement="center"></ic-divider>
        <ic-divider label="Label right" label-placement="right"></ic-divider>
        <ic-divider label-placement="left">
          <ic-typography variant="label" slot="label">
            <p style="width: max-content;">Slotted label left</p>
          </ic-typography></ic-divider
        >
        <ic-divider label-placement="center">
          <ic-typography variant="label" slot="label">
            <p style="width: max-content;">Slotted label center</p>
          </ic-typography></ic-divider
        >
        <ic-divider label-placement="right">
          <ic-typography variant="label" slot="label">
            <p style="width: max-content;">Slotted label right</p>
          </ic-typography></ic-divider
        >
      </div>
      <div
        style="display: flex; gap: var(--ic-space-lg); height: 200px; margin-top: var(--ic-space-lg)"
      >
        <ic-divider
          label="Label top"
          label-placement="top"
          orientation="vertical"
        ></ic-divider>
        <ic-divider
          label="Label center"
          label-placement="center"
          orientation="vertical"
        ></ic-divider>
        <ic-divider
          label="Label bottom"
          label-placement="bottom"
          orientation="vertical"
        ></ic-divider>
        <ic-divider label-placement="top" orientation="vertical">
          <ic-typography variant="label" slot="label">
            <p style="width: max-content;">Slotted label top</p>
          </ic-typography></ic-divider
        >
        <ic-divider label-placement="center" orientation="vertical">
          <ic-typography variant="label" slot="label">
            <p style="width: max-content;">Slotted label center</p>
          </ic-typography></ic-divider
        >
        <ic-divider label-placement="bottom" orientation="vertical">
          <ic-typography variant="label" slot="label">
            <p style="width: max-content;">Slotted label bottom</p>
          </ic-typography></ic-divider
        >
      </div>`}
  </Story>
</Canvas>

### Playground

export const defaultArgs = {
  "border-style": "solid",
  label: "",
  "label-placement": "none",
  monochrome: false,
  orientation: "horizontal",
  theme: "inherit",
  weight: "thin",
  "--ic-divider-horizontal-width": "100%",
  "--ic-divider-vertical-height": "400px",
  "--ic-divider-background": "",
  "--ic-divider-background-monochrome": "",
  "--ic-divider-label": "",
  "--ic-divider-label-monochrome": "",
  "--ic-divider-background": "",
  "--ic-divider-label-width": "",
};

<Canvas>
  <Story
    args={defaultArgs}
    parameters={{ loki: { skip: true } }}
    argTypes={{
      "border-style": {
        options: ["solid", "dashed"],
        control: {
          type: "inline-radio",
        },
      },
      label: {
        mapping: {
          none: "",
        },
      },
      "label-placement": {
        options: ["none", "left", "center", "right", "top", "bottom"],
        mapping: {
          none: "",
        },
        control: {
          type: "radio",
        },
      },
      orientation: {
        options: ["horizontal", "vertical"],
        control: {
          type: "inline-radio",
        },
      },
      theme: {
        options: ["inherit", "dark", "light"],
        control: {
          type: "radio",
        },
      },
      weight: {
        options: ["thin", "medium", "thick", "very-thick"],
        control: {
          type: "radio",
        },
      },
      "--ic-divider-background": {
        control: {
          type: "color",
        },
      },
      "--ic-divider-background-monochrome": {
        control: {
          type: "color",
        },
      },
      "--ic-divider-label": {
        control: {
          type: "color",
        },
      },
      "--ic-divider-label-monochrome": {
        control: {
          type: "color",
        },
      },
    }}
    name="Playground"
  >
    {(args) => {
      const backgroundColor =
        args.theme === "dark"
          ? "var(--ic-color-background-primary-dark)"
          : args.theme === "light"
          ? "var(--ic-color-background-primary-light)"
          : "transparent";
      return html`<div
        style="display: flex; width: 100%; height: 100%; padding: var(--ic-space-md); background-color: ${backgroundColor}; box-sizing: border-box;"
      >
        <ic-divider
          monochrome=${args.monochrome}
          orientation=${args.orientation}
          theme=${args.theme}
          border-style=${args["border-style"]}
          weight=${args.weight}
          label=${args.label}
          label-placement=${args["label-placement"]}
          style="
            --ic-divider-horizontal-width: ${args[
            "--ic-divider-horizontal-width"
          ]};
            --ic-divider-vertical-height: ${args[
            "--ic-divider-vertical-height"
          ]};
            ${args["--ic-divider-background"] !== ""
            ? `--ic-divider-background: ${args["--ic-divider-background"]};`
            : ""}
            ${args["--ic-divider-background-monochrome"] !== ""
            ? `--ic-divider-background-monochrome: ${args["--ic-divider-background-monochrome"]};`
            : ""}
            ${args["--ic-divider-label"] !== ""
            ? `--ic-divider-label: ${args["--ic-divider-label"]};`
            : ""}
            ${args["--ic-divider-label-monochrome"] !== ""
            ? `--ic-divider-label-monochrome: ${args["--ic-divider-label-monochrome"]};`
            : ""}
            ${args["--ic-divider-label-width"] !== ""
            ? `--ic-divider-label-width: ${args["--ic-divider-label-width"]};`
            : ""}"
        ></ic-divider>
      </div>`;
    }}
  </Story>
</Canvas>
