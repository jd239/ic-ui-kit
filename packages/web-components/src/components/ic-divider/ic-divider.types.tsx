export type IcDividerStyles = "solid" | "dashed";
export type IcDividerWeights = "thin" | "medium" | "thick" | "very-thick";
export type IcDividerLabelPlacement =
  | "left"
  | "center"
  | "right"
  | "top"
  | "bottom";
