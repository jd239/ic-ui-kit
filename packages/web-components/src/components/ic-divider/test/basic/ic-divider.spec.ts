import { newSpecPage } from "@stencil/core/testing";
import { Divider } from "../../ic-divider";
import { Typography } from "../../../ic-typography/ic-typography";

describe("ic-divider", () => {
  it("should render", async () => {
    const page = await newSpecPage({
      components: [Divider],
      html: `<ic-divider></ic-divider>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render with a slotted label", async () => {
    const page = await newSpecPage({
      components: [Divider],
      html: `<ic-divider label-placement="left">
        <ic-typography variant="label" slot="label">
          <p style="width: max-content;">Slotted label left</p>
        </ic-typography>
      </ic-divider>`,
    });

    expect(page.root).toMatchSnapshot();
  });

  it("should render a label with the label typography variant when weight is not set to thick or very-thick and a label is provided", async () => {
    const page = await newSpecPage({
      components: [Divider, Typography],
      html: `<ic-divider weight="medium" label="Test label"></ic-divider>`,
    });

    const label = page.root.querySelector("ic-typography");

    expect(label.variant).toBe("label");

    expect(page.root).toMatchSnapshot();
  });

  it("should render a label with the subtitle-large typography variant when weight is set to thick and a label is provided", async () => {
    const page = await newSpecPage({
      components: [Divider, Typography],
      html: `<ic-divider weight="thick" label="Test label"></ic-divider>`,
    });

    const label = page.root.querySelector("ic-typography");

    expect(label.variant).toBe("subtitle-large");

    expect(page.root).toMatchSnapshot();
  });

  it("should render a label with the h4 typography variant when weight is set to very-thick and a label is provided", async () => {
    const page = await newSpecPage({
      components: [Divider, Typography],
      html: `<ic-divider weight="very-thick" label="Test label"></ic-divider>`,
    });

    const label = page.root.querySelector("ic-typography");

    expect(label.variant).toBe("h4");

    expect(page.root).toMatchSnapshot();
  });

  it("should console error if there's an invalid prop combination of the placement prop and the orientation", async () => {
    console.error = jest.fn();
    const page = await newSpecPage({
      components: [Divider],
      html: `<ic-divider orientation="vertical" label-placement="left" label="Test label"></ic-divider>`,
    });

    await page.waitForChanges();

    expect(console.error).toHaveBeenCalledWith(
      "Left label placement is not applicable for a vertical ic-divider (web-components) / IcDivider (react)"
    );
  });

  it("correctly sets foregroundColor on theme change", async () => {
    const page = await newSpecPage({
      components: [Divider],
      html: `<ic-side-navigation version="v0.0.0" status="BETA" app-title="ACME">
      <ic-navigation-item slot="primary-navigation" href="/" label="Home">
            <svg
              slot="icon"
              width="24"
              height="24"
              viewBox="0 0 24 24"
              fill="none"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M12 6.19L17 10.69V18.5H15V12.5H9V18.5H7V10.69L12 6.19ZM12 3.5L2 12.5H5V20.5H11V14.5H13V20.5H19V12.5H22L12 3.5Z"
                fill="currentColor"
              />
            </svg>
          </ic-navigation-item>
      <ic-divider slot="primary-navigation"></ic-divider>
        <ic-navigation-item slot="primary-navigation" href="/" label="a11y">
          <svg
            slot="icon"
            width="24"
            height="24"
            viewBox="0 0 24 24"
            fill="none"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 2C13.1 2 14 2.9 14 4C14 5.1 13.1 6 12 6C10.9 6 10 5.1 10 4C10 2.9 10.9 2 12 2ZM21 9H15V22H13V16H11V22H9V9H3V7H21V9Z"
              fill="currentColor"
            />
          </svg>
        </ic-navigation-item>
        </ic-side-navigation>`,
    });

    expect(page.rootInstance.foregroundColor).toEqual("light");
    expect(page.rootInstance.theme).toEqual("dark");

    expect(page.root).toMatchSnapshot("theme-change-light");

    await page.rootInstance.themeChangeHandler({
      detail: { mode: "dark" },
    });

    await page.rootInstance.updateMonochromeState();
    await page.waitForChanges();

    expect(page.rootInstance.foregroundColor).toEqual("dark");
    expect(page.rootInstance.theme).toEqual("light");

    expect(page.root).toMatchSnapshot("theme-change-dark");
  });
});
