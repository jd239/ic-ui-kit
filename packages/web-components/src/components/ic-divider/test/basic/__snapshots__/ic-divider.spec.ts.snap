// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ic-divider correctly sets foregroundColor on theme change: theme-change-dark 1`] = `
<ic-divider class="ic-divider-horizontal ic-divider-solid ic-divider-thin ic-side-navigation-keyline ic-theme-light" slot="primary-navigation">
  <!---->
  <hr>
</ic-divider>
`;

exports[`ic-divider correctly sets foregroundColor on theme change: theme-change-light 1`] = `
<ic-divider class="ic-divider-horizontal ic-divider-solid ic-divider-thin ic-side-navigation-keyline ic-theme-dark" slot="primary-navigation">
  <!---->
  <hr>
</ic-divider>
`;

exports[`ic-divider should render 1`] = `
<ic-divider class="ic-divider-horizontal ic-divider-solid ic-divider-thin">
  <!---->
  <hr>
</ic-divider>
`;

exports[`ic-divider should render a label with the h4 typography variant when weight is set to very-thick and a label is provided 1`] = `
<ic-divider aria-orientation="horizontal" class="ic-divider-horizontal ic-divider-solid ic-divider-very-thick" label="Test label" role="separator" weight="very-thick">
  <!---->
  <ic-typography class="ic-divider-label ic-typography-h4">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    <p>
      Test label
    </p>
  </ic-typography>
</ic-divider>
`;

exports[`ic-divider should render a label with the label typography variant when weight is not set to thick or very-thick and a label is provided 1`] = `
<ic-divider aria-orientation="horizontal" class="ic-divider-horizontal ic-divider-medium ic-divider-solid" label="Test label" role="separator" weight="medium">
  <!---->
  <ic-typography class="ic-divider-label ic-typography-label">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    <p>
      Test label
    </p>
  </ic-typography>
</ic-divider>
`;

exports[`ic-divider should render a label with the subtitle-large typography variant when weight is set to thick and a label is provided 1`] = `
<ic-divider aria-orientation="horizontal" class="ic-divider-horizontal ic-divider-solid ic-divider-thick" label="Test label" role="separator" weight="thick">
  <!---->
  <ic-typography class="ic-divider-label ic-typography-subtitle-large">
    <mock:shadow-root>
      <slot></slot>
    </mock:shadow-root>
    <p>
      Test label
    </p>
  </ic-typography>
</ic-divider>
`;

exports[`ic-divider should render with a slotted label 1`] = `
<ic-divider aria-orientation="horizontal" class="ic-divider-horizontal ic-divider-solid ic-divider-thin" label-placement="left" role="separator">
  <!---->
  <div class="ic-divider-label-container">
    <ic-typography slot="label" variant="label">
      <p style="width: max-content;">
        Slotted label left
      </p>
    </ic-typography>
  </div>
</ic-divider>
`;
